### ISSUE TITLE
How to setup "ODE_weights" and “data_weights” ？

### ISSUE THREAD

**Issue by NibofWar** [lu-group/sbinn#6]
Hi ,Thank you for your contribyution to this work.
When I run a inverse ODE problem，I am confused about the setting of ODE weights. What rules or experience do you use to determine ODE_weights and data_weights?

**Comment by mitchelldaneker** [comment]
Hello,

There are a few methods that can be used to set the weights. The one we used is one of the easier methods to implement. Essentially, we start by setting all the weights to one and calculate the initial loss (at zero epochs, this can be done by just running and stopping after the loss at zero epochs is calculated). Then, we set the weights such that when we run it again, all portions of the loss are about the same order of magnitude at zero epochs. We have found this helps with training. 

For example, let's say your loss output has 2 PDE functions and some observations. So the loss would be of the form [pde1, pde2, observation]. At zero epochs, if the loss was [11.5, .00345, 0.25], a way to weight would be [0.1,1000,10]. This would result in a loss of approximately [1.5, 3.45, 2.5] at zero epochs when you run again. Note that it will not be the same, due to random initialization. 

Ideally, this method should work or at least get you very close to where you need. It's possible the weights will have to be slightly adjusted, and this is part of the tuning process of the network. 

Some other methods you can look into are those that will update the loss weights as the network trains. These can be implemented in DeepXDE via callbacks and I encourage you to check out the DeepXDE documentation if you want to use one of these methods. 

Let me know if you have any questions. Also, with the latest version of DeepXDE, epochs is depreciated and iterations is the new term that is used.

**Comment by chenyv118** [comment]
> 

Hello, loss weights do have a considerable impact on the training results, I tried many different weight combinations, your method does work, but I still can't get the right results, I went to look for some adaptive weight tuning solutions, but they more or less need to tweak the loss function or the network structure, can you please guide me:
1. does Deepxde integrate methods for adaptive tuning of loss weights? I can't find such a method.
2. does Deepxde support designing network structure and loss function by oneself? Where can I see such sample code for my reference?

**Comment by mitchelldaneker** [comment]
Currently, DeepXDE does not integrate any methods for adaptive tuning of loss weights. If you would like to alter the network structure, what is currently available is FNN, and PFNN, but not recombination's of layers in PFNN ([16,16,[16,16]] is allowed but [16,[16,16],16] is not). You may define your own loss function, which can be observed via the FAQ posted below for your convenience. 

Q: Implement new losses or constraints.
A: [174](https://github.com/lululxvi/deepxde/issues/174), [286](https://github.com/lululxvi/deepxde/issues/286), [311](https://github.com/lululxvi/deepxde/issues/311), [356](https://github.com/lululxvi/deepxde/issues/356), [460](https://github.com/lululxvi/deepxde/issues/460), [467](https://github.com/lululxvi/deepxde/issues/467), [504](https://github.com/lululxvi/deepxde/issues/504)

If you want to go beyond these bounds, you may have to alter the DeepXDE code. You can also use early stopping, and change the weights ever couple of iterations.

**Comment by chenyv118** [comment]
> Currently, DeepXDE does not integrate any methods for adaptive tuning of loss weights. If you would like to alter the network structure, what is currently available is FNN, and PFNN, but not recombination's of layers in PFNN ([16,16,[16,16]] is allowed but [16,[16,16],16] is not). You may define your own loss function, which can be observed via the FAQ posted below for your convenience.
> 
> Q: Implement new losses or constraints. A: [174](https://github.com/lululxvi/deepxde/issues/174), [286](https://github.com/lululxvi/deepxde/issues/286), [311](https://github.com/lululxvi/deepxde/issues/311), [356](https://github.com/lululxvi/deepxde/issues/356), [460](https://github.com/lululxvi/deepxde/issues/460), [467](https://github.com/lululxvi/deepxde/issues/467), [504](https://github.com/lululxvi/deepxde/issues/504)
> 
> If you want to go beyond these bounds, you may have to alter the DeepXDE code. You can also use early stopping, and change the weights ever couple of iterations.

Thank you for your reply, I will continue my work based on the information you provided.

**Comment by chenyv118** [comment]
> Currently, DeepXDE does not integrate any methods for adaptive tuning of loss weights. If you would like to alter the network structure, what is currently available is FNN, and PFNN, but not recombination's of layers in PFNN ([16,16,[16,16]] is allowed but [16,[16,16],16] is not). You may define your own loss function, which can be observed via the FAQ posted below for your convenience.
> 
> Q: Implement new losses or constraints. A: [174](https://github.com/lululxvi/deepxde/issues/174), [286](https://github.com/lululxvi/deepxde/issues/286), [311](https://github.com/lululxvi/deepxde/issues/311), [356](https://github.com/lululxvi/deepxde/issues/356), [460](https://github.com/lululxvi/deepxde/issues/460), [467](https://github.com/lululxvi/deepxde/issues/467), [504](https://github.com/lululxvi/deepxde/issues/504)
> 
> If you want to go beyond these bounds, you may have to alter the DeepXDE code. You can also use early stopping, and change the weights ever couple of iterations.

I am sorry to bother you again, in your reply I missed a point before, I need to use early stopping if I wish to adjust the loss weights, I didn't understand the usage of this early stopping, if I need to change the loss weights of the model I need to stop the training and recompile it right? I tried to use the callback function to adjust the loss weight every few epochs, when I debugged it I could see that model.losshistory.loss_weight did change, but the loss value during training remained unchanged.

Here is my implementation of the callback function:
```python
import torch
from deepxde.callbacks import Callback
from softadapt import SoftAdapt


class AdaptLossWeight(Callback):
    def __init__(self, num_component, period=1):
        super().__init__()
        self.period = period
        self.num_component = num_component

        self.cnt = 0
        self.loss_history = [[] for _ in range(num_component)]
        self.loss_adapt = SoftAdapt(beta=-1)

    def on_epoch_end(self):
        if self.cnt < self.period:
            self.cnt += 1
            for i in range(self.num_component):
                self.loss_history[i].append(self.model.train_state.loss_train[i])
            return
        self.cnt = 0

        loss_component_values = [torch.Tensor(self.loss_history[i]) for i in range(self.num_component)]
        
        # This is the adjusted loss weight
        adapt_weight = self.loss_adapt.get_component_weights(
            *loss_component_values, verbose=False
        )
        self.model.losshistory.set_loss_weights(adapt_weight)
        print(self.model.losshistory.loss_weights)
        self.loss_history = [[] for _ in range(40)]

```

**Comment by mitchelldaneker** [comment]
It may be easier to just recompile the model every time you want to change the weights. Compiling is fast so it should not add too much time to your run. I am not entirely familiar with everything you have called, if `self.model.losshistory.set_loss_weights()` only looks at the loss history, and not the training losses, that would be the reason you are running into the current issues. Let me know if this works, I am off my vacation now so will be able to respond more promptly.

**Comment by chenyv118** [comment]
> It may be easier to just recompile the model every time you want to change the weights. Compiling is fast so it should not add too much time to your run. I am not entirely familiar with everything you have called, if `self.model.losshistory.set_loss_weights()` only looks at the loss history, and not the training losses, that would be the reason you are running into the current issues. Let me know if this works, I am off my vacation now so will be able to respond more promptly.

Yes you are right, I followed what you said and recompiled the model every time I needed to change the weights and it did work, thank you very much for your help! (Although the adaptive weighting method I tried still doesn't solve the problem)

**Comment by mitchelldaneker** [comment]
Not sure what exactly your problem is you are trying to solve, but often normalizing variables and doing a change of variable in the ODE can go a long way.

**Comment by chenyv118** [comment]
> Not sure what exactly your problem is you are trying to solve, but often normalizing variables and doing a change of variable in the ODE can go a long way.

OK, I will always remember your suggestion with sincere thanks.

**Comment by mitchelldaneker** [comment]
Due to inactivity on this issue I will close it. Please reopen if you need more assistance.
